### PR TITLE
pkg/trace: Update stats payload format

### PR DESCRIPTION
**Pull request by piochelepiotr**
This PR moves the agent to use a new stats payload format.
The reasons are:
- Start using DDSketch for stats for more accurate percentiles
- Move away from the json encoding we use today and to reduce the payload size
- More efficient payload format to reduce CPU / memory impact of agent & backend

### CODE REVIEW

File: pkg/trace/agent/agent.go [hunk 1]
```diff
@@ -150,7 +147,10 @@ func (a *Agent) loop() {
 
 // Process is the default work unit that receives a trace, transforms it and
 // passes it downstream.
-func (a *Agent) Process(p *api.Payload, sublayerCalculator *stats.SublayerCalculator) {
+func (a *Agent) Process(p *api.Payload) {
+	if p.ClientComputedStats {
```
Comment (furmmon) [inline pkg/trace/agent/agent.go:151]: this would prevent any sampling

File: pkg/trace/agent/agent_test.go [hunk 2]
```diff
@@ -328,7 +327,7 @@ func TestProcess(t *testing.T) {
 		go agnt.Process(&api.Payload{
 			Traces: traces,
 			Source: agnt.Receiver.Stats.GetTagStats(info.Tags{}),
-		}, stats.NewSublayerCalculator())
+		})
```
Comment (furmmon) [inline pkg/trace/agent/agent_test.go:330]: I think it would be better to drop sublayers in another PR, on which you would base stats computation.
It makes it harder to read and it's actually worth a release note on its own

File: pkg/trace/agent/agent.go (hunk 1 above)
Comment (piochelepiotr) [reply pkg/trace/agent/agent.go]: Right

File: pkg/trace/agent/agent_test.go (hunk 2 above)
Comment (piochelepiotr) [reply pkg/trace/agent/agent_test.go:326]: Makes sense.

File: pkg/trace/agent/agent.go [hunk 3]
```diff
@@ -150,15 +147,18 @@ func (a *Agent) loop() {
...
 		return
 	}
 	defer timing.Since("datadog.trace_agent.internal.process_payload_ms", time.Now())
 	ts := p.Source
 	ss := new(writer.SampledSpans)
-	sinputs := make([]stats.Input, 0, len(p.Traces))
+	var sinputs []stats.Input
+	if !p.ClientComputedStats {
+		sinputs = make([]stats.Input, 0, len(p.Traces))
+	}
```
Comment (gbbr) [inline pkg/trace/agent/agent.go:161]: I think you can completely drop this. It is fine to append to a slice which hasn't been "made"

Comment (furmmon) [reply pkg/trace/agent/agent.go]: I think it shouldn't be dropped: it will generate a lot of reallocations: trace payloads can have many traces and slices are reallocated by x2 (2,4,8,16 ... 
The cost of this allocating this array at the right size is already visible in some of my profiles

Comment (gbbr) [reply pkg/trace/agent/agent.go]: That's true, good point. In that case maybe we can move it into the other `if !p.ClientComputedStats {` clause. No need to have them so far away from each other. It's inviting bugs :)

File: pkg/trace/writer/stats.go [hunk 4]
```diff
@@ -16,45 +18,43 @@ import (
...
+	"github.com/tinylib/msgp/msgp"
 )
 
 // pathStats is the target host API path for delivering stats.
 const pathStats = "/api/v0.2/stats"
 
 const (
 	// bytesPerEntry specifies the approximate size an entry in a stat payload occupies.
-	bytesPerEntry = 125
+	bytesPerEntry = 375
```
Comment (piochelepiotr) [inline pkg/trace/writer/stats.go:32]: we need to do real tests to get these metrics right.

File: pkg/trace/stats/aggregation.go [hunk 5]
```diff
@@ -21,107 +22,40 @@ type Aggregation struct {
...
+	return uint32(c)
+}
+
 // NewAggregationFromSpan creates a new aggregation from the provided span and env
-func NewAggregationFromSpan(s *pb.Span, env string) Aggregation {
+func NewAggregationFromSpan(s *pb.Span, env string, defaultHostname string) Aggregation {
 	synthetics := strings.HasPrefix(s.Meta[tagOrigin], "synthetics")
-
+	hostname := s.Meta[tagHostname]
+	if hostname == "" {
```
Comment (gbbr) [inline pkg/trace/stats/aggregation.go:48]: Maybe it can never happen, but it crossed my mind so leaving a comment. Let's use `GetMeta` here to avoid any surprises :)
```suggestion
	hostname := defaultHostname
	if v, ok := traceutil.GetMeta(s, tagHostname); ok {
		hostname = v
	}
```

File: pkg/trace/stats/concentrator.go [hunk 6]
```diff
@@ -32,19 +33,18 @@ type Concentrator struct {
...
-
-	exit   chan struct{}
-	exitWG *sync.WaitGroup
-
-	buckets map[int64]*RawBucket // buckets used to aggregate stats per timestamp
-	mu      sync.Mutex
+	In        chan []Input
+	Out       chan []*pb.ClientStatsPayload
+	exit      chan struct{}
+	exitWG    *sync.WaitGroup
```
Comment (gbbr) [inline pkg/trace/stats/concentrator.go:39]: I know this isn't your work but if you removed this pointer it would be so nice :) It's not even necessary to initialise waitgroups

**Review by gbbr — APPROVED** [review]
This LGTM in terms of code and approach. Quite great actually 🙂

**Comment by mx-psi** [comment]
Hi, is this still targeting 7.27? Do you plan to merge it soon? I am going to do a PR for moving some things around on the Trace Agent to create a Go module and I would like to avoid causing a merge conflict for either of us

**Comment by piochelepiotr** [comment]
> Hi, is this still targeting 7.27? Do you plan to merge it soon? I am going to open a PR soonish for moving some things around on the Trace Agent to create a Go module and I would like to avoid causing a merge conflict for either of us

Hi! Yes, I hope I will get it merged tomorrow.

**Comment by mx-psi** [comment]
Awesome, thanks for the reply!

File: pkg/trace/stats/aggregation.go (hunk 5 above)
Comment (piochelepiotr) [reply pkg/trace/stats/aggregation.go:48]: Sounds good. I added a method: `GetMetaDefault`

File: pkg/trace/stats/concentrator.go [hunk 7]
```diff
@@ -6,9 +6,11 @@
 package stats
 
 import (
+	"github.com/DataDog/datadog-agent/pkg/trace/config"
```
Comment (gbbr) [inline pkg/trace/stats/concentrator.go:9]: Please make sure to move these imports down with the right group in all files (I see it multiple places)

File: pkg/trace/stats/statsraw.go [hunk 8]
```diff
@@ -8,26 +8,63 @@ package stats
...
+	"github.com/DataDog/datadog-agent/pkg/trace/pb"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+
+	"github.com/DataDog/sketches-go/ddsketch"
+	"github.com/golang/protobuf/proto"
+)
+
+const (
+	relativeAccuracy = 0.01
+	maxNumBins       = 2048
```
Comment (gbbr) [inline pkg/trace/stats/statsraw.go:20]: Make sure to explain these.

File: pkg/trace/stats/statsraw.go [hunk 9]
```diff
@@ -52,8 +89,15 @@ type RawBucket struct {
 	keyBuf strings.Builder
 }
 
+// PayloadKey is the key of a stats payload.
```
Comment (gbbr) [inline pkg/trace/stats/statsraw.go:92]: This isn't really documentation 😛

**Review by gbbr — COMMENTED** [review]
Left a few nits, but don't mind them, I can see you're pushing forward to get this merged before the freeze. That's fine. But please follow up with a PR which improves documentation 🙂

File: pkg/trace/writer/stats.go [hunk 10]
```diff
@@ -129,35 +131,23 @@ func (w *StatsWriter) Stop() {
...
-		"Content-Type":     "application/json",
+		"Content-Type":     "application/msgpack",
 		"Content-Encoding": "gzip",
 	})
-	if err := stats.EncodePayload(req.body, p); err != nil {
+	if err := encodePayload(req.body, p); err != nil {
 		log.Errorf("Stats encoding error: %v", err)
 		return
 	}
-	atomic.AddInt64(&w.stats.Bytes, int64(req.body.Len()))
```
Comment (piochelepiotr) [inline pkg/trace/writer/stats.go:159]: we add that when we sent the payload already. We were double counting.

File: pkg/trace/stats/statsraw.go (hunk 9 above)
Comment (piochelepiotr) [reply pkg/trace/stats/statsraw.go:92]: 😆 true, actually, I don't need to export that anymore.

File: pkg/trace/agent/agent_test.go [hunk 11]
```diff
@@ -366,82 +366,6 @@ func TestProcess(t *testing.T) {
 	})
 }
 
-func TestFilteredByTags(t *testing.T) {
```
Comment (bmermet) [inline pkg/trace/agent/agent_test.go:369]: Why was this test removed? `filteredByTags` is not removed by this PR.

File: pkg/trace/stats/aggregation.go [hunk 12]
```diff
@@ -21,107 +24,41 @@ type Aggregation struct {
...
-		Version:    s.Meta[tagVersion],
-		Synthetics: synthetics,
+func getStatusCode(s *pb.Span) uint32 {
+	strC := traceutil.GetMetaDefault(s, tagStatusCode, "")
+	if strC == "" {
+		return 0
+	}
+	c, err := strconv.Atoi(strC)
+	if err != nil {
+		log.Errorf("Invalid status code %s. Using 0.", strC)
```
Comment (bmermet) [inline pkg/trace/stats/aggregation.go:41]: Is it the right level for an agent log? If the instrumentation is misconfigured this will flood the logs. From what I can see in other place most logging for invalid data is done with `log.Debugf`. @gbbr will know better than me.

File: pkg/trace/stats/statsraw.go [hunk 13]
```diff
@@ -8,26 +8,68 @@ package stats
...
+	// 80 micro second to 1 year: http://www.vldb.org/pvldb/vol12/p2195-masson.pdf
+	maxNumBins = 2048
 )
 
 // Most "algorithm" stuff here is tested with stats_test.go as what is important
 // is that the final data, the one with send after a call to Export(), is correct.
 
 type groupedStats struct {
-	topLevel float64
+	hits            float64
```
Comment (bmermet) [inline pkg/trace/stats/statsraw.go:32]: What's the reason why the counts are float64 here but uint64 in the proto?

File: releasenotes/notes/apm-new-stats-payload-e718e56797cd408f.yaml [hunk 14]
```diff
@@ -0,0 +1,12 @@
...
+# CHANGELOG.rst is rendered. So the text needs to be worded so that
+# it does not depend on any information only available in another
+# section. This may mean repeating some details, but each section
+# must be readable independently of the other.
+#
+# Each section note must be formatted as reStructuredText.
+---
+enhancements:
+  - |
+    APM: Change in the Stats payload format leading to reduced CPU and memory usage.
```
Comment (bmermet) [inline releasenotes/notes/apm-new-stats-payload-e718e56797cd408f.yaml:11]: nit: why is stats spelled with a capital S here?

**Review by raphaelgavache — APPROVED** [review]
lgtm

File: pkg/trace/stats/aggregation.go (hunk 12 above)
Comment (furmmon) [reply pkg/trace/stats/aggregation.go:41]: This is actually a redundant check as it's done before by the normalizer: https://github.com/DataDog/datadog-agent/blob/master/pkg/trace/agent/normalizer.go#L117-L122

File: pkg/trace/agent/agent_test.go (hunk 11 above)
Comment (piochelepiotr) [reply pkg/trace/agent/agent_test.go:369]: Good catch, added it back.

File: pkg/trace/stats/statsraw.go (hunk 13 above)
Comment (piochelepiotr) [reply pkg/trace/stats/statsraw.go]: No good reason, I just didn't change it. Now it makes a lot more sense indeed. Updated.

File: pkg/trace/writer/stats.go [hunk 15]
```diff
@@ -168,112 +158,143 @@ func (w *StatsWriter) sendPayloads() {
...
-func (w *StatsWriter) buildPayloads(s []stats.Bucket, maxEntriesPerPayloads int) ([]*stats.Payload, int, int) {
-	if len(s) == 0 {
-		return []*stats.Payload{}, 0, 0
+// encodePayload encodes the payload as Gzipped msgPack into w.
+func encodePayload(w io.Writer, payload pb.StatsPayload) error {
+	gz, err := gzip.NewWriterLevel(w, gzip.BestSpeed)
+	if err != nil {
+		return err
+	}
+	_, err = gz.Write([]byte{v1})
```
Comment (bmermet) [inline pkg/trace/writer/stats.go:167]: This is probably too late to change that, but it's an odd way of versioning a API. Usually we version the api path. If I understand correctly the payload format is no longer valid msgpack since there is an extra byte.

File: pkg/trace/writer/stats.go [hunk 16]
```diff
@@ -168,112 +158,143 @@ func (w *StatsWriter) sendPayloads() {
...
+			},
+		}
 	}
 	// 3. Iterate over all entries of each stats. Add the entry to one of
-	//    the payload container mappings, in a round robin fashion. In some
-	//    edge cases, we can end up having the same entry in several
-	//    inputted stat buckets. We must check that we never overwrite an
-	//    entry in the new stats buckets but cleanly merge instead.
+	//    the payloads, in a round robin fashion. Use the indexes maps to
+	//    ensure that we have one ClientStatsBucket per timeWindow for each ClientStatsPayoad.
```
Comment (bmermet) [inline pkg/trace/writer/stats.go:270]: ClientStatsPayoad -> ClientStatsPayload

File: pkg/trace/stats/concentrator.go [hunk 17]
```diff
@@ -31,34 +33,33 @@ type Concentrator struct {
...
-	In  chan []Input
-	Out chan []Bucket
-
-	exit   chan struct{}
-	exitWG *sync.WaitGroup
-
-	buckets map[int64]*RawBucket // buckets used to aggregate stats per timestamp
-	mu      sync.Mutex
+	bufferLen     int
+	In            chan []Input
```
Comment (gbbr) [inline pkg/trace/stats/concentrator.go:37]: If you've merged all of this together, let's create two groups:
* exported fields at the top, then one empty line and
* unexported fields

File: pkg/trace/stats/statsraw.go [hunk 18]
```diff
@@ -6,28 +6,82 @@
...
+		Duration:       round(s.duration),
+		TopLevelHits:   round(s.topLevelHits),
+		OkSummary:      okSummary,
+		ErrorSummary:   errSummary,
+		Synthetics:     k.aggr.Synthetics,
+	}, nil
 }
 
 func newGroupedStats() *groupedStats {
+	ok, _ := ddsketch.LogCollapsingLowestDenseDDSketch(relativeAccuracy, maxNumBins)
```
Comment (gbbr) [inline pkg/trace/stats/statsraw.go:80]: Is it safe to ignore the second return value?

File: pkg/trace/stats/statsraw.go [hunk 19]
```diff
@@ -52,8 +106,15 @@ type RawBucket struct {
 	keyBuf strings.Builder
 }
 
+// PayloadKey uniquely identifies a ClientStatsPayload inside a StatsPayload
```
Comment (gbbr) [inline pkg/trace/stats/statsraw.go:109]: Much better thank you :)

File: pkg/trace/test/testutil/stats.go [hunk 20]
```diff
@@ -6,22 +6,35 @@
...
 	srb := stats.NewRawBucket(0, 1e9)
 	for _, s := range spans {
-		srb.HandleSpan(s, defaultEnv)
+		// override version to ensure all buckets will have the same payload key.
+		s.Meta["version"] = ""
+		srb.HandleSpan(s, defaultEnv, defaultHostname)
+	}
+	buckets := srb.Export()
+	if len(buckets) != 1 {
+		panic("All entries must have the same payload key.")
```
Comment (gbbr) [inline pkg/trace/test/testutil/stats.go:28]: This seems risky 🙂 Do we really want to have a full crash and panic here?

File: pkg/trace/writer/stats.go [hunk 21]
```diff
@@ -168,112 +158,143 @@ func (w *StatsWriter) sendPayloads() {
...
+// and then groups them into pb.StatsPayload with less than maxEntriesPerPayload
+func (w *StatsWriter) buildPayloads(sp pb.StatsPayload, maxEntriesPerPayload int) []pb.StatsPayload {
+	split := splitPayloads(sp.Stats, maxEntriesPerPayload)
+	grouped := make([]pb.StatsPayload, 0, len(sp.Stats))
+	current := pb.StatsPayload{
+		AgentHostname: sp.AgentHostname,
+		AgentEnv:      sp.AgentEnv,
+	}
+	nbEntries := 0
+	nbBuckets := 0
```
Comment (gbbr) [inline pkg/trace/writer/stats.go:189]: var nbEntries, nbBuckets int

File: pkg/trace/test/testutil/stats.go (hunk 20 above)
Comment (piochelepiotr) [reply pkg/trace/test/testutil/stats.go:28]: This is in `testutil`, do you think it's an issue to panic during tests?

File: pkg/trace/stats/statsraw.go (hunk 18 above)
Comment (piochelepiotr) [reply pkg/trace/stats/statsraw.go:80]: The error is only when relativeAccuracy is outside of [0,1] interval, and it's a constant here.
I'm a bit hesitant to add an error path for an error that can't happen.
What would you do in that case?

File: pkg/trace/test/testutil/stats.go (hunk 20 above)
Comment (gbbr) [reply pkg/trace/test/testutil/stats.go:28]: You're right. Sorry I missed that. Resolving.

File: pkg/trace/stats/statsraw.go (hunk 18 above)
Comment (gbbr) [reply pkg/trace/stats/statsraw.go:80]: How about simply `log.Error` and that's it?

Comment (piochelepiotr) [reply pkg/trace/stats/statsraw.go]: Sounds good. Updated.

File: pkg/trace/writer/stats.go (hunk 15 above)
Comment (piochelepiotr) [reply pkg/trace/writer/stats.go]: I removed the version bit.

**Review by bmermet — APPROVED** [review]

**Review by gbbr — APPROVED** [review]
Really nice work Piotr! This is a fantastic improvement and cleanup!
